### ISSUE TITLE
tensorflow training: result is different if model passed as argument to training loop

### ISSUE THREAD

**Issue by animalcroc** [tensorflow/tensorflow#55189]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes and no. the training loops are stock but the rest of code is not
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): microsoft windows 10 pro
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.1
- Python version: 3.7.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: none
- GPU model and memory: NVIDIA QUATRO RTX 4000 with max-q design

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I have experimented with training a simple tensorflow model using two scenarios: passing my model to my training loop (and to the subfunctions which are called from training loop), versus not passing my model to the training loop. The two cases result in different results. When passing my model to the training functions, the model is trained properly. But in the second scenario, something is wrong because the model is apparently not trained. I am baffled, and I wonder if it's a scope thing.

To be more specific, my setup involves dynamically creating a new model of larger size (adding some layers at each iteration of a for loop), and then training the resulting model. As stated before, I train the model in two scenarios: passing the model to training subfunctions and not doing so, and I obtain different results depending on which one I do. I verify this by giving the model a test sample (class 0 MNIST images) and checking if the correct classification is output. The models trained by passing the model as an argument is trained correctly, but, if I do not do this, then only the first model created by the for loop is correctly trained, as verified by incorrect class predictions. Can this be explained?

**Describe the expected behavior**
Training a model should not be different whether or not I pass the model as an argument to the custom training function.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
no

- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

`
import tensorflow as tf
physical_devices = tf.config.experimental.list_physical_devices('GPU')
tf.config.experimental.set_memory_growth(physical_devices[0], True)
from tensorflow.keras.optimizers import Adam
from tensorflow import keras
from tensorflow.keras import layers
import numpy as np
import time

epochs = 200
input_shape = (28,28,1)
num_classes=10
batch_size = 64
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

x_val = x_train[-10000:]
y_val = y_train[-10000:]
x_train = x_train[:-10000]
y_train = y_train[:-10000]

x_train = np.expand_dims(x_train, -1)
x_val = np.expand_dims(x_val, -1)
x_test = np.expand_dims(x_test, -1)
x_train = x_train.astype("float32") 
x_test = x_test.astype("float32") 
x_val = x_val.astype("float32") 

y_test_sorted_args_0=np.where(y_test == 0)
x_test_0=x_test[y_test_sorted_args_0]
y_test_0=np.full( (x_test_0.shape)[0],0)

train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)

val_dataset = tf.data.Dataset.from_tensor_slices((x_val, y_val))
val_dataset = val_dataset.batch(batch_size)

acc_metric = keras.metrics.SparseCategoricalAccuracy()
val_acc_metric = keras.metrics.SparseCategoricalAccuracy()
optimizer = keras.optimizers.SGD(learning_rate=1e-3)
loss_fn = keras.losses.SparseCategoricalCrossentropy(from_logits=False)
    
@tf.function
def train_step(x, y):
    with tf.GradientTape() as tape:
        mod_output = model(x, training=True)
        loss_value = loss_fn(y, mod_output)
    grads = tape.gradient(loss_value, model.trainable_weights)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
    acc_metric.update_state(y, mod_output)
    return loss_value

@tf.function
def test_step(x, y):
    val = model(x, training=False)
    acc_metric.update_state(y, val)

def train( epochs):
    for epoch in range(epochs):
        print("\nStart of epoch %d" % (epoch,))
        start_time = time.time()
        
        # Iterate over the batches of the dataset.
        for step, (x_batch_train, y_batch_train) in enumerate(train_dataset):
            loss_value = train_step( x_batch_train, y_batch_train)
    
            # Log every 200 batches.
            if step % 200 == 0:
                print(
                    "Training loss (for one batch) at step %d: %.4f"
                    % (step, float(loss_value))
                )
                print("Seen so far: %d samples" % ((step + 1) * batch_size))
    
        # Display metrics at the end of each epoch.
        train_acc = acc_metric.result()
        print("Training acc over epoch: %.4f" % (float(train_acc),))
    
        # Reset training metrics at the end of each epoch
        acc_metric.reset_states()
    
        # Run a validation loop at the end of each epoch.
        for x_batch_val, y_batch_val in val_dataset:
            test_step(x_batch_val, y_batch_val)
    
        val_acc = acc_metric.result()
        val_acc_metric.reset_states()
        print("Validation acc: %.4f" % (float(val_acc),))
        print("Time taken: %.2fs" % (time.time() - start_time))

max_hidden=7
for num_hidden_layers in range(1,max_hidden,3):    
    model1 = keras.Sequential(
        [
            keras.Input(shape=input_shape),
            layers.Flatten(),
        ]
    )
    
    for i in range(1, num_hidden_layers+1):
        model1.add(layers.Dense(150, activation="relu"))
    model1.add(layers.Dense(num_classes, activation="softmax"))
    
    model=model1
    
    train(epochs)
    
    #verify that the model is properly trained by checking that the model correclty predicts images from class 0.  
    #The more class 0 predictions we have, the better.
    for sample_index in range(0,10): 
        x_sample=x_test_0[sample_index,:,:,:]
        x_sample=np.expand_dims(x_sample, 0)
        print(tf.math.argmax(model(x_sample),axis=1))
    time.sleep(1)
        
`

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by animalcroc** [comment]
For some reason I cannot properly add my code to this log. Some of it is correctly recognized and put in gray, but a lot of it is not

**Comment by sushreebarsa** [comment]
@animalcroc  We see that you are using older version of TF (2.1) which is not actively supported.We recommend you to kindly upgrade to 2.4 or later versions and let us know if the issue still persists ?Thanks!

**Comment by animalcroc** [comment]
I will update you soon

**Comment by sushreebarsa** [comment]
@animalcroc Could you please let us know if there is any update on this issue?
Thanks!

**Comment by animalcroc** [comment]
Yes, I will tomorrow. Thanks for checking with me.

**Comment by animalcroc** [comment]
Yes, I will tomorrow. Thanks for checking with me.

**Comment by sushreebarsa** [comment]
@animalcroc 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by animalcroc** [comment]
The code was included in my original post. Here it is again:

`import tensorflow as tf
physical_devices = tf.config.experimental.list_physical_devices('GPU')
tf.config.experimental.set_memory_growth(physical_devices[0], True)
from tensorflow.keras.optimizers import Adam
from tensorflow import keras
from tensorflow.keras import layers
import numpy as np
import time

epochs = 200
input_shape = (28,28,1)
num_classes=10
batch_size = 64
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

x_val = x_train[-10000:]
y_val = y_train[-10000:]
x_train = x_train[:-10000]
y_train = y_train[:-10000]

x_train = np.expand_dims(x_train, -1)
x_val = np.expand_dims(x_val, -1)
x_test = np.expand_dims(x_test, -1)
x_train = x_train.astype("float32")
x_test = x_test.astype("float32")
x_val = x_val.astype("float32")

y_test_sorted_args_0=np.where(y_test == 0)
x_test_0=x_test[y_test_sorted_args_0]
y_test_0=np.full( (x_test_0.shape)[0],0)

train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)

val_dataset = tf.data.Dataset.from_tensor_slices((x_val, y_val))
val_dataset = val_dataset.batch(batch_size)

acc_metric = keras.metrics.SparseCategoricalAccuracy()
val_acc_metric = keras.metrics.SparseCategoricalAccuracy()
optimizer = keras.optimizers.SGD(learning_rate=1e-3)
loss_fn = keras.losses.SparseCategoricalCrossentropy(from_logits=False)

@tf.function
def train_step(x, y):
  with tf.GradientTape() as tape:
  mod_output = model(x, training=True)
  loss_value = loss_fn(y, mod_output)
  grads = tape.gradient(loss_value, model.trainable_weights)
  optimizer.apply_gradients(zip(grads, model.trainable_weights))
  acc_metric.update_state(y, mod_output)
return loss_value

@tf.function
  def test_step(x, y):
  val = model(x, training=False)
  acc_metric.update_state(y, val)

def train( epochs):
  for epoch in range(epochs):
  print("\nStart of epoch %d" % (epoch,))
  start_time = time.time()

# Iterate over the batches of the dataset.
  for step, (x_batch_train, y_batch_train) in enumerate(train_dataset):
        loss_value = train_step( x_batch_train, y_batch_train)

        # Log every 200 batches.
        if step % 200 == 0:
            print(
                "Training loss (for one batch) at step %d: %.4f"
                % (step, float(loss_value))
            )
            print("Seen so far: %d samples" % ((step + 1) * batch_size))

# Display metrics at the end of each epoch.
  train_acc = acc_metric.result()
  print("Training acc over epoch: %.4f" % (float(train_acc),))

    # Reset training metrics at the end of each epoch
  acc_metric.reset_states()    

    # Run a validation loop at the end of each epoch.
  for x_batch_val, y_batch_val in val_dataset:
        test_step(x_batch_val, y_batch_val)

  val_acc = acc_metric.result()
  val_acc_metric.reset_states()
  print("Validation acc: %.4f" % (float(val_acc),))
  print("Time taken: %.2fs" % (time.time() - start_time))

  max_hidden=7
  for num_hidden_layers in range(1,max_hidden,3):
    model1 = keras.Sequential(
      [
      keras.Input(shape=input_shape),
      layers.Flatten(),
      ])

  for i in range(1, num_hidden_layers+1):
    model1.add(layers.Dense(150, activation="relu"))
  model1.add(layers.Dense(num_classes, activation="softmax"))

  model=model1

  train(epochs)

#verify that the model is properly trained by checking that the model correclty predicts images from class 0.  
#The more class 0 predictions we have, the better.
  for sample_index in range(0,10): 
    x_sample=x_test_0[sample_index,:,:,:]
    x_sample=np.expand_dims(x_sample, 0)
    print(tf.math.argmax(model(x_sample),axis=1))
  time.sleep(1)
`

**Comment by animalcroc** [comment]
I included code in my original post, but the formatting is not correct. Only part of the code is gray, and the rest is in text format. I can't fix the formatting even though I try

**Comment by sushreebarsa** [comment]
@animalcroc 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!

**Comment by animalcroc** [comment]
I just posted there. Here is the link: https://github.com/keras-team/keras/issues/16240

**Comment by sushreebarsa** [comment]
@animalcroc Thank you for the confirmation!
Can you please move this issue to closed status as we will track the other ticket there.
Thanks!

**Comment by animalcroc** [comment]
Closed per request from sushreebarsa)
